### ISSUE TITLE
Unforking possible?

### ISSUE THREAD

**Issue by timofonic** [joncampbell123/dosbox-x#327]
Hello.
Is it possible "unfork" of some or them or at least merge interesting changes?

Thanks a lot in advance!

- @Henne [Bright Eyes](https://github.com/Henne/Bright-Eyes)
- [dosbox](https://github.com/svn2github/dosbox)
- @duganchen's [DosBox](https://github.com/duganchen/dosbox)
- @libretro's [dosbox-libretro](https://github.com/libretro/dosbox-libretro/)
(maybe there's others. I'm not able to check it right now...)

**Comment by duganchen** [comment]
The "interesting changes" from my fork are available as a series of patches, here:

http://www.vogons.org/viewtopic.php?f=41&t=43068&start=20#p568163

**Comment by joncampbell123** [comment]
Well, the two codebases are similar but DOSBox-X has diverged quite a bit design-wise and organization-wise.

I suggest porting each feature one at a time and carefully.

**Comment by sikthehedgehog** [comment]
> "Bright Eyes" is a variation of DOSBox with customized support for the CRPG "Das schwarze Auge 1 - Schicksalsklinge"

What does the Bright Eyes fork do exactly? I can't tell if it's adding game-specific stuff or just implementing bugfixes. Also would be nice to know how that game works on DOSBox-X.

**Comment by joncampbell123** [comment]
Looking at the commits and source code, Bright Eyes appears to be DOSBox with in-emulator patching and hacks for a specific game.

I'd be OK with code *from* DOSBox-X making it's way into Bright Eyes but I cannot accept game-specific hacks into DOSBox-X. DOSBox-X is about accuracy in emulation. So if a game has programming mistakes and rendering glitches on real hardware then it must have the same in DOSBox-X.

**Comment by duganchen** [comment]
If we want a list of DosBox forks with interesting features: one feature I'd really like to see as a patchset is Grid Cartographer integration.

http://gridcartographer.com/download/

**Comment by joncampbell123** [comment]
I'm going to have to decline integrating game-specific editing into DOSBox-X. However, I would have no problem with a fork of DOSBox-X that wants to focus on game-specific hacks or editing.

I have considered that it would be useful to have a hook or intercept system for debugging and game-hacking purposes, though. It could be done in a general manner so that you could then load whatever debugging/game-hacking code you like. It would include VRAM access to an "overlay" layer atop VGA emulation so your 1337 hacking code could show status or information and support for hooking specific keys on the keyboard. I've often thought how neat it would be if you did this by running an EXE file of your choice from a hidden "system management mode" environment running atop the DOS environment.

**Comment by timofonic** [comment]
@joncampbell123 Yes, that intercept system would be very interesting to have. Add that and scripting support by using Lua or other(s) and game hacks would be less invasive and even have their own cureated repository section for people to use them :)

FCEUX uses Lua scripting...
http://www.fceux.com/web/help/fceux.html?LuaScripting.html
https://github.com/BillyWM/FCEUX-Lua-Script-Collection
https://github.com/nodchip/fceux-lua-simulated-annealing
https://github.com/haseeb-heaven/LuaScripts_NES
https://github.com/rdcox/FCEUX-scripting
https://github.com/haseeb-heaven/LuaRio_Bot
https://github.com/anthonyduggan/fceux-livesplit-socket
https://github.com/anthonyduggan/SMB1Debug

[BizHawk](https://github.com/TASVideos/BizHawk) supports scripting too.

I think you may get lots of feedback from projects like [ScummVM](https://github.com/scummvm/scummvm) and maybe others, but this one is quite active and even have an irc channel (#scummvm at irc.freenode.net). They are using to advanced forms of debugging (reverse engineering), so I think they are the best at specifying what would make their efforts easier and better for them or future contributors :)

**Comment by joncampbell123** [comment]
Lua would be fine as well, though it would require yet another addon to the VS2015 project files.

I only recommended the "System Management Mode" method because it would allow re-using the existing x86 decoder and execution code.

**Comment by timofonic** [comment]
@joncampbell123 Anyway, it would be amazing if you get able to get feedback from projects like ScummVM to get an idea about what to do. Please consider it :D

**Comment by joncampbell123** [comment]
Will do. This is the place to comment on it.

**Comment by timofonic** [comment]
There's other forks...
- @mypasswordisqwerty's [dosbox-python](/mypasswordisqwerty/dosbox-python)
- @wjp's [Dosbox](/wjp/dosbox) fork contains [idados](/wjp/idados) code.
- @lab313ru's [idados_dosbox](/lab313ru/idados_dosbox)
- @pinchyCZN's [dosbox](https://github.com/pinchyCZN/dosbox): Custom build to debug mpxplay.
- @wwiv @wwivbbs [dosbox](/wwiv/dosbox) fork for BBS stuff.
- @mspicer [dosbox_log-parse](/mspicer/dosbox_log-parse): Python script to parse a Dosbox CPU debugging log to make it easier to find information you need.

Blabla: http://www.vogons.org/viewtopic.php?f=32&t=7323&sid=1d1105c57812bcb66a362d116c9102f1&start=80

**Comment by sikthehedgehog** [comment]
If I were given a vote, I'd really prefer to avoid including all the application-specific stuff in DOSBox-X... Would prefer it to focus on improving emulation accuracy than trying to cram in just about every shiny thing everybody made x_x

If *really* insisting, probably the best would be for some way to allow scripting that can be used to build all those features on (though they'll need to be redone from scratch), but putting all that into the core emulator itself seems like a disaster waiting to happen in the long term.

Also let's not forget that whenever a feature is added, it needs to be maintained forever :P

**Comment by timofonic** [comment]
@sikthehedgehog I think some things are subjective, it depends on the personal or group interests.

I agree about making application-specific stuff as less and easy to do as possible. This would be quite interesting for people interesting in gaming hacks and such, those could be great DosBox-X betatesters in a coincidental way too.

@joncampbell123 Did you talk with ScummVM Team? It's very easy, they have an IRC channel so just write and wait until they reply (they are from different parts of the world and not all are available at same time). Those people use advanced tools and might help at providing feedback to make your project more useful for debugging and a lot more. Maybe you can get their interest and receive patches. There are other projects that reimplement DOS-era game engines that might be interested on certain features too :)

**Comment by joncampbell123** [comment]
@timofonic I'll consider it, but that's not my focus right now.

**Comment by wwiv** [comment]
My fork isn't much of a fork.  I needed something newer than 0.74 since that release broke the serial port code for working with BBS doors.  Also needed the locking patch added to work with BBS doors as well. 

Eventually I'll get around to deleting things that aren't needed, but don't want to make it harder to keep up to date from SVN.  Also I've tried a SDL2 patch to no success, would like to do that too, just to stay up to date on the dependencies.  I've also had to patch it to work around the SDL1 screen redraw problems under vmware.

**Comment by duganchen** [comment]
TBH, I’d like to do a new port to SDL2 one day. The existing one suffers
greatly from being primarily a port to Android.

The only reason I haven’t done so is because I’m busy with something else
these days.

**Comment by sikthehedgehog** [comment]
> TBH, I’d like to do a new port to SDL2 one day. The existing one suffers greatly from being primarily a port to Android.

It's also completely unbuildable for me right now because SDL1 from the Ubuntu repo is broken (some obscure dependency deep down is broken) and I can't get it to build (I get a compiler error). So huh, at the moment I can't use DOSBox-X at all until it's ported to SDL2.

Pretty sure there's an existing issue number for porting to SDL2.

**Comment by joncampbell123** [comment]
@sikthehedgehog I've modified the code so that you can run ./configure to select whether you to compile against SDL 1.x or SDL 2.x.

Due to SDL 1.x's deprecated status, I'll eventually pull the source code into DOSBox-X's source tree and maintain it as the SDL 1.x support until SDL 1.x is eventually deprecated entirely in DOSBox-X.

**Comment by joncampbell123** [comment]
Just a future idea for SDL 2.x builds and touchscreens:

If scriptable hacking is allowed in DOSBox-X, a good use of it would be to allow the script to take in touchscreen events and generate them to the guest however it sees fit as a way of adapting old games to touchscreen input.

**Comment by joncampbell123** [comment]
Also I've pulled SDL 1.x into the DOSBox-X source tree and I've already made some fixes to the code to improve SDL 1.x builds, such as one issue with Windows 10 and mouse motion tracking that caused random large jumps in the input.

**Comment by timofonic** [comment]
@joncampbell123 Okay, no problem. I hope someday you consider it. Thanks a lot.
